### ISSUE TITLE
`self` arguments to methods should not have type annotations

### ISSUE THREAD

**Issue by rra** [astral-sh/ruff#4396]
Currently, Ruff does not appear to exclude the `self` argument to class methods from its check for type annotations. For example, this Python file:

```python
class Foo:
    def bar(self, something: str) -> str:
        return something + "else"
```

reports:

```
% ruff check --select 'ANN' foo.py 
foo.py:2:13: ANN101 Missing type annotation for `self` in method
Found 1 error.
```

However, `self` should not have a type annotation; it's unnecessary (it adds no useful type information; the type of self is already known by the type checker because the method is defined on a class) and explicitly recommended against in the mypy documentation. See, for example, https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html#classes (note the comment on `deposit`).

**Comment by charliermarsh** [comment]
I believe `ANN101` and `ANN102` are their own rules for this reason -- so that you can turn them off. Whether they should exist _at all_ is probably a different question, but the encoding was inherited from the originating plugin.

**Comment by rra** [comment]
Ah, I see this is controversial upstream as well (https://github.com/sco1/flake8-annotations/issues/80).

I think `ANN101` is always incorrect and indeed would like the linter to check the inverse: complain if `self` has a type annotation, since the type annotation for `self` makes the code more fragile under refactoring and should be omitted. But I totally understand if you don't want to diverge from the original upstream plugin, and it doesn't look like the plugin author is interested in changing this.

**Comment by madkinsz** [comment]
Now that there's a [`Self` type](https://peps.python.org/pep-0673/) it's a bit more reasonable to expect a type annotation here for consistency with other function parameters.

**Comment by rra** [comment]
Incidentally, flake8-annotations links to https://peps.python.org/pep-0484/#annotating-instance-and-class-methods as an example of when this typing might be necessary, but I believe this example is now obsolete now that `Self` has been added. The example in PEP 484:
```python
T = TypeVar('T', bound='Copyable')
class Copyable:
    def copy(self: T) -> T:
        # return a copy of self
```
can now be written much simpler as:
```
class Copyable:
    def copy(self) -> Self:
        # Return a copy of self
```

**Comment by rra** [comment]
> Now that there's a [`Self` type](https://peps.python.org/pep-0673/) it's a bit more reasonable to expect a type annotation here for consistency with other function parameters.

Agreed that at least makes it less fragile and turns it into a question of style about which people will have different opinions. I modify my desire for a linter for `self` typing to one that complains about any type other than `Self`. :)

**Comment by charliermarsh** [comment]
Ah interesting. Yeah I've definitely done this before for generics:

```py
T = TypeVar('T', bound='Copyable')
class Copyable:
    def copy(self: T) -> T:
        # return a copy of self
```

It's neat that it's no longer required.

**Comment by charliermarsh** [comment]
I think we need to do some grepping to see if any projects are actually relying on this rule, or whether we can just remove it.

**Comment by bersbersbers** [comment]
Duplicate of #7709, I guess

**Comment by charliermarsh** [comment]
Ah that's right! This rule is removed in v0.2.0.

**Comment by ssbarnea** [comment]
Does anyone know an reverse-ANN001 that can be used to remove mistakenly added type to self?
